add DataFrame.count_rows() to get length without materialization (#358)

Asrst opened the pull request:
Implemented changes to close #358

- Added `DataFrame.count_rows()` methods which return count of dataframe rows without materialization (Integer).
- Added a `LocalCount` logical plan to return length of each partition. `count_rows()` chains `LocalCount`+ `Coalesce(1)`+ `LocalAggregate(sum)`  
-  Implemented necessary handles & physical plans for LocalCount to work with both PyRunner & DynamicRunner.
- `len(df)` returns (`DataFrame.__len__`) length when dataframe materialized using `df.collect()`
- `len(df)` will throw an error on an unmaterialized DataFrame.
- Tests not implemented yet (will open another Issue/PR with below test plan)

**Edit:** Test's are added into `tests/dataframe_cookbook/test_count_rows.py`

```python

import datetime
from daft import DataFrame


df = DataFrame.from_pydict({
    "A": [1, 1, 3, 4],
    "B": [1.5, 2.5, 3.5, 4.5],
    "C": [True, True, False, False],
    "D": ["a", "b", "c", "d"],
    "E": [b"a", b"b", b"c", b"d"],
    "F": [datetime.date(1994, 1, 1), datetime.date(1994, 1, 2), datetime.date(1994, 1, 3), datetime.date(1994, 1, 4)],
    "G": [[1, 1, 1], [2, 2, 2], [3, 3, 3], [4, 4, 4]],
})

# count_rows (returns 4)
print(df.count_rows()) 

# reparition & count (returns 4)
num_rows = df.repartition(3).count_rows()
print(num_rows)

# group by & count (returns 3)
print(df.groupby('A').sum('B').count_rows())

# return error message that df is not materialized
len(df)

# return length of dataframe
df.collect()
print(len(df))



```

jaychia commented on daft/execution/execution_step.py:
```diff
@@ -230,6 +230,27 @@ def run(self, inputs: list[vPartition]) -> list[vPartition]:
+    def _count(self, inputs: list[vPartition]) -> list[vPartition]:
+        [input] = inputs
+
+        partition = LocalLogicalPartitionOpRunner()._handle_local_count(
```
Once we clean up code in `_handle_local_count` to use `vPartition.count_rows()`, we can use that `vPartition.count_rows()` API here directly as well, which should clean up this `LocalCount` instruction quite a bit.

jaychia commented on daft/execution/physical_plan_factory.py:
```diff
@@ -59,6 +59,9 @@ def get_physical_plan(node: LogicalPlan) -> Iterator[None | ExecutionStep[Partit
             )
+        
+        elif isinstance(node, logical_plan.LocalCount):
+            return physical_plan.local_count(child_plan, node)
```
I believe local count is actually a pipelineable instruction. Once we do the refactors to use `vPartition.count_rows()`, we should be able to do:

```
return physical_plan.pipeline_instruction(
    child_plan=child_plan,
    pipeable_instruction=execution_step.LocalCount(),
)
```

jaychia commented on daft/logical/logical_plan.py:
```diff
@@ -593,6 +593,32 @@ def rebuild(self) -> LogicalPlan:
 
+class LocalCount(UnaryNode):
+    def __init__(self, input: LogicalPlan) -> None:
+        schema = Schema([Field("count", ExpressionType.from_py_type(int))])
```
```suggestion
        schema = Schema([Field("count", ExpressionType.integer())])
```

jaychia commented on daft/dataframe/dataframe.py:
```diff
@@ -1188,6 +1211,21 @@ def collect(self, num_preview_rows: int | None = 10) -> DataFrame:
+            return len(self._result)
+
+        message = "No data loaded. Call `df.collect()` to materialize the Dataframe first."
+        raise Exception(message)
```
Nit: let's raise a RuntimeError instead?

```
raise RuntimeError("Cannot call len() on an unmaterialized dataframe: either materialize your dataframe with df.collect() first before calling len(), or use `df.count_rows()` instead which will calculate the total number of rows")
```

jaychia commented on daft/execution/logical_op_runners.py:
```diff
@@ -296,7 +323,7 @@ def _handle_in_memory_scan(self, inputs: dict[int, PartitionSet], im_scan: InMem
         return result
 
-    def _handle_global_limit(self, inputs: dict[int, PartitionSet], limit: GlobalLimit) -> PartitionSet:
+    def _handle_global_limit(self, inputs: dict[int, PartitionSet], limit: GlobalLimit):
```
```suggestion
    def _handle_global_limit(self, inputs: dict[int, PartitionSet], limit: GlobalLimit) -> PartitionSet:
```

jaychia commented on daft/logical/logical_plan.py:
```diff
@@ -593,6 +593,32 @@ def rebuild(self) -> LogicalPlan:
+    def __repr__(self) -> str:
+        return self._repr_helper()
+
+    def resource_request(self) -> ResourceRequest:
```
You can just remove this now! We made a refactor recently that returns `None` for the resource_request by default on the LogicalPlan super type

jaychia reviewed (commented):
Looking pretty good!

Some nits, but would really love it if you could move the counting kernel into `vPartition`. This will help us move this logic into Rust soon, and will clean up our code quite a bit as well!

Asrst commented:
thanks @jaychia . the suggestions make sense. Will make neccesary changes

Asrst commented:
- Refactored the code to include suggested changes.
- Improved local count logic by using `vPartition.__len__()` and `vPartition.from_pydict()`.
- `len(daft_df)` will now raises a runtime error if Dataframe is  unmaterialized, instead of an exception.
- Added tests (`test_count_rows.py`) to validate both `DataFrame.count_rows()` and `DataFrame.__len__()`

jaychia commented on daft/dataframe/dataframe.py:
```diff
@@ -802,6 +802,23 @@ def limit(self, num: int) -> DataFrame:
+    
+    @DataframePublicAPI
+    def count_rows(self) -> int:
+        """Performs a LocalCount() + Coalesce(1) + LocalAggregate(sum) to get number of rows.
```
```suggestion
        """Executes the Dataframe to count the number of rows
```

jaychia commented on daft/logical/logical_plan.py:
```diff
@@ -593,6 +593,29 @@ def rebuild(self) -> LogicalPlan:
+        return LocalCount(new_children[0])
+
+    def required_columns(self) -> set[str]:
+        return set()
```
This is technically true, but actually might mess with the way our optimizer works. Our optimizer might see this and think "none of the columns are required, so no work needs to be done".

Instead, let's do the naive thing and just grab the first column from the child plan for now?

We can add a comment here explaining why this is done.

```suggestion
        # HACK: Arbitrarily return the first column in the child to ensure that
        # at least one column is computed by the optimizer
        return self._children()[0].schema().column_names()[0]
```

jaychia commented:
Looking great, love the tests!

Added a few more suggestions, and enabled the PR to run CI checks

Asrst commented:
My working branch is behind and tests are failing here as result of those changes. I will fetch the changes & will fix these issues. 
Thanks!

Asrst commented:
hi @jaychia, 

I am returning the python integer data type by using `result.value.get_partition(0).to_pydict()['count'][0]` in the `DataFrame.count_rows()`. However, this is failing on the ray runners with error: `ray._raylet.ObjectRef' object has no attribute 'to_pydict'`. Is this expected behaviour on ray runners?

is there a common way/method to get python dict from both `LocalPartitionSet` & `RayPartitionSet`?

jaychia commented:
> hi @jaychia,
> 
> I am returning the python integer data type by using `result.value.get_partition(0).to_pydict()['count'][0]` in the `DataFrame.count_rows()`. However, this is failing on the ray runners with error: `ray._raylet.ObjectRef' object has no attribute 'to_pydict'`. Is this expected behaviour on ray runners?
> 
> is there a common way/method to get python dict from both `LocalPartitionSet` & `RayPartitionSet`?
> 
> **Edit:** Figured the reason for the error. I am not waiting for async result. Found a better way to get the result using `num_rows = DataFrame(local_sum_op).to_pydict()['count'][0]`

Yes you got it! In the Ray Runner, our Partition objects are actually Ray `ObjectRef`s, which are "pointers" to a partition that is in distributed memory on the Ray cluster. Thus you can't actually call `.to_pydict()` on this pointer - you'd have to write a Ray remote function to do so.

`num_rows = DataFrame(local_sum_op).to_pydict()['count'][0]` is perfect - it does all this under the hood for you :)

jaychia commented on daft/logical/logical_plan.py:
```diff
@@ -584,6 +584,31 @@ def rebuild(self) -> LogicalPlan:
+    def required_columns(self) -> set[str]:
+        # HACK: Arbitrarily return the first column in the child to ensure that
+        # at least one column is computed by the optimizer
+        return self._children()[0].schema().column_names()[0]
```
My apologies - I think my code was wrong! This needs to return a `set`

```suggestion
        return {self._children()[0].schema().column_names()[0]}
```

jaychia commented:
Rest of the PR looks good! Let us know in Discord if you need help getting all the style stuff correct - it should be pretty easy to run locally. Looks like it's failing on a `mypy` issue that was from my previous code suggestion 🤯 - try my new code suggestion out and see if it works.

jaychia reviewed (approved):
Looks great! Amazing work @Asrst :)

jaychia commented:
Congrats on your first PR @Asrst ! 🥳 🥳 🥳

samster25 commented:
Thanks for all the hard work! @Asrst
